keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1619 @@
+the `resource_id` has already been prepared, the Plugin MUST reply `0
+OK`.
+
+If this RPC failed, or kubelet does not know if it failed or not, it
+MAY choose to call `NodePrepareResource` again, or choose to call
+`NodeUnprepareResource`.
+
+After a successful call, Kubelet MUST pass device names and kinds to
+the runtime through the CRI protocol.
+
+```
+<<[UNRESOLVED @bart0sh]>>
+CRI protocol may need to be extended for
+this purpose, e.g. device id can be added to the CRI Device structure.
+<<[/UNRESOLVED]>>
+```
+
+```protobuf
+message NodePrepareResourceRequest {
+  // The UID of the ResourceClaim. This field is REQUIRED.
+  string resource_uid = 1;
+}
+
+message NodePrepareResourceResponse {
+  // These are the additional devices that kubelet must
+  // make available via the container runtime. A resource
+  // may have zero or more devices.
+  repeated CDIDevice devices = 1;
+}
+
+message CDIDevice {
+  // Kind is the string that together with the name identifies a device
+  // (https://github.com/container-orchestrated-devices/container-device-interface/blob/master/SPEC.md#kind).
+  string kind = 1;
+  // Name is the name that within its kind uniquely identifies a
+  // device (https://github.com/container-orchestrated-devices/container-device-interface/blob/master/SPEC.md#cdi-devices).
+  string name = 2;
+}
+```
+
+###### NodePrepareResource Errors
+
+If the plugin is unable to complete the NodePrepareResource call
+successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST
+return the specified gRPC error code.  Kublet MUST implement the
+specified error recovery behavior when it encounters the gRPC error
+code.
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-----------|-----------|-------------|-------------------|
+| Resource does not exist | 5 NOT_FOUND | Indicates that a resource corresponding to the specified `resource_id` does not exist. | Caller MUST verify that the `resource_id` is correct and that the resource is accessible and has not been deleted before retrying with exponential back off. |
+ 
+
+##### `NodeUnprepareResource`
+
+A Kubelet Plugin MUST implement this RPC call. This RPC is a reverse
+operation of `NodePrepareResource`. This RPC MUST undo the work by
+the corresponding `NodePrepareResource`. This RPC SHALL be called by
+kubelet at least once for each successful `NodePrepareResource`. The
+Plugin SHALL assume that this RPC will be executed on the node where
+the resource is being used.
+
+This RPC is called by kubelet when the Pod using the resource is being
+deleted.
+
+This operation MUST be idempotent. If this RPC failed, or kubelet does
+not know if it failed or not, it can choose to call
+`NodeUnprepareResource` again.
+
+```protobuf
+message NodeUnprepareResourceRequest {
+  // The UID of the ResourceClaim. This field is REQUIRED.
+  string resource_id = 1;
```

Comment:
Would it make sense to add CDI device ids here so kubelet plugin don't need to keep resource_id -> CDI device ids map?

Reply:
That smells like a shortcut to me that might be too limiting for some resource drivers.

For example, can it happen that different resource allocations use the same device id but additional changes are done outside of the container runtime for each of them? Then NodeUnprepareResourceRequest wouldn't know which resource is meant.

I suppose we can send both (`resource_id` and the device ids). kubelet must have both and the overhead is irrelevant.

Reply:
> I suppose we can send both (resource_id and the device ids).

that's what I proposed - to *add* device ids.

Reply:
I misread your initial sentence ("replace" instead of "add"). Care to provide a PR with that change?

Reply:
sure, will do